server/src/grpc/remote_engine_service/mod.rs

```diff
@@ -581,28 +613,51 @@ impl RemoteEngineServiceImpl {
         request: Request<ExecutePlanRequest>,
     ) -> Result<StreamWithMetric<ExecutePlanMetricCollector>> {
         let metric = ExecutePlanMetricCollector(Instant::now());
+        let (tx, rx) = mpsc::channel(STREAM_QUERY_CHANNEL_LEN);
         let request = request.into_inner();
         let query_engine = self.instance.query_engine.clone();
         let (ctx, encoded_plan) = extract_plan_from_req(request)?;
 
-        let stream = self
+        let physical_plan = Arc::new(DataFusionPhysicalPlanAdapter::new(TypedPlan::Remote(
+            encoded_plan,
+        )));
+        let physical_plan_clone = physical_plan.clone();
+
+        let mut stream = self
             .runtimes
             .read_runtime
-            .spawn(async move { handle_execute_plan(ctx, encoded_plan, query_engine).await })
+            .spawn(async move { handle_execute_plan(ctx, physical_plan_clone, query_engine).await })
             .await
             .box_err()
             .with_context(|| ErrWithCause {
                 code: StatusCode::Internal,
                 msg: "failed to run execute physical plan task",
-            })??
-            .map(|result| {
-                result.box_err().context(ErrWithCause {
-                    code: StatusCode::Internal,
-                    msg: "failed to poll record batch for remote physical plan",
-                })
-            });
+            })??;
+
+        self.runtimes.read_runtime.spawn(async move {
+            while let Some(result) = stream.next().await {
+                let result = result
+                    .map(RecordBatchWithMetrics::RecordBatch)
+                    .box_err()
+                    .context(ErrWithCause {
+                        code: StatusCode::Internal,
+                        msg: "failed to poll record batch for remote physical plan",
+                    });
+                if let Err(e) = tx.send(result).await {
+                    error!("Failed to send handler result, err:{}.", e);
+                    break;
+                }
+            }
+            let metrics = physical_plan.metrics_to_string();
+            if let Err(e) = tx.send(Ok(RecordBatchWithMetrics::Metrics(metrics))).await {
+                error!("Failed to send handler result, err:{}.", e);
+            }
+        });
 
-        Ok(StreamWithMetric::new(Box::pin(stream), metric))
+        Ok(StreamWithMetric::new(
+            Box::pin(ReceiverStream::new(rx)),
+            metric,
+        ))
```

Comment:
Should not use channel to do that... It introduces too many costs to  this path.
I think we just need to define a stream wrapper here, when poll the noraml query path to end, we get and return the metric before returning None.